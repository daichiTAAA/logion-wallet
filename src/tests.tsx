import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };
