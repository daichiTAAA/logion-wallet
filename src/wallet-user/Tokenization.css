.Tokenization {
    margin-bottom: 30px;
}
