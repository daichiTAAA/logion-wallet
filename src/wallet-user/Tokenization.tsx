import React, { useState, useCallback } from "react";
import { useForm } from 'react-hook-form';

import { useCommonContext } from "../common/CommonContext";
import Button from "../common/Button";
import Dialog from "../common/Dialog";
import { DEFAULT_LEGAL_OFFICER } from '../common/types/LegalOfficer';

import { useUserContext } from './UserContext';
import { CreateTokenRequest } from "./Model";
import TokenCreationForm from './TokenCreationForm';

import './Tokenization.css';

export enum State {
    START,
    REQUEST_TOKENIZATION
}

export interface Props {
    initialState?: State
}

interface FormValues {
    requestedTokenName: string,
    bars: number,
}

export default function Tokenization(props: Props) {
    const { accounts, colorTheme } = useCommonContext();
    const { refreshRequests, createTokenRequest } = useUserContext();
    const [state, setState] = useState(props.initialState !== null ? props.initialState : State.START)
    const { control, handleSubmit, formState: { errors }, reset } = useForm<FormValues>();

    const submit = useCallback((formValues: FormValues) => {
        (async function() {
            const currentAddress = accounts!.current!.address;
            const request: CreateTokenRequest = {
                legalOfficerAddress: DEFAULT_LEGAL_OFFICER,
                requesterAddress: currentAddress,
                bars: Number(formValues.bars),
                requestedTokenName: formValues.requestedTokenName,
            }
            await createTokenRequest!(request);
            reset();
            setState(State.START);
            refreshRequests!(true);
        })();
    }, [ accounts, createTokenRequest, setState, refreshRequests, reset ]);

    return (
        <div className="Tokenization">
            <Button onClick={ () => setState(State.REQUEST_TOKENIZATION) }>
                Request token creation
            </Button>
            <Dialog
                show={ state === State.REQUEST_TOKENIZATION }
                size="lg"
                actions={[
                    {
                        id: "submit",
                        buttonText: 'Submit',
                        buttonVariant: 'primary',
                        type: 'submit',
                    },
                    {
                        id: "cancel",
                        callback: () => { reset() ; setState(State.START) },
                        buttonText: 'Cancel',
                        buttonVariant: 'secondary',
                    }
                ]}
                onSubmit={handleSubmit(submit)}
            >
                <TokenCreationForm
                    control={ control }
                    errors={ errors }
                    colors={ colorTheme.dialog }
                />
            </Dialog>
        </div>
    );
}
