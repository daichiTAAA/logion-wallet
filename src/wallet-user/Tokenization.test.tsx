import React from "react";
import {shallowRender} from "../tests";
import Tokenization, {State} from "./Tokenization";

test("START: Enable button", () => {
    const tree = shallowRender(<Tokenization/>)
    expect(tree).toMatchSnapshot();
});

test("REQUEST_TOKENIZATION: Disable button, Show form", () => {
    const tree = shallowRender(<Tokenization initialState={State.REQUEST_TOKENIZATION}/>)
    expect(tree).toMatchSnapshot();
});
test("REQUEST_TOKENIZATION_DONE: Enable button, Show message", () => {
    const tree = shallowRender(<Tokenization initialState={State.REQUEST_TOKENIZATION_DONE}/>)
    expect(tree).toMatchSnapshot();
});
