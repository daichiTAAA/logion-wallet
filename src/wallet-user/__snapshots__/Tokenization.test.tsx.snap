// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REQUEST_TOKENIZATION: Disable button, Show form 1`] = `
<div
  className="Tokenization"
>
  <Button
    onClick={[Function]}
  >
    Request token creation
  </Button>
  <Dialog
    actions={
      Array [
        Object {
          "buttonText": "Submit",
          "buttonVariant": "primary",
          "id": "submit",
          "type": "submit",
        },
        Object {
          "buttonText": "Cancel",
          "buttonVariant": "secondary",
          "callback": [Function],
          "id": "cancel",
        },
      ]
    }
    onSubmit={[Function]}
    show={true}
    size="lg"
  >
    <TokenCreationForm
      colors={
        Object {
          "background": "#eff3fe",
          "borderColor": "#e6007a",
          "foreground": "#000000",
        }
      }
      control={
        Object {
          "defaultValuesRef": Object {
            "current": Object {},
          },
          "fieldArrayDefaultValuesRef": Object {
            "current": Object {},
          },
          "fieldsRef": Object {
            "current": Object {},
          },
          "formStateRef": Object {
            "current": Object {
              "dirtyFields": Object {},
              "errors": Object {},
              "isDirty": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": false,
              "isValidating": false,
              "submitCount": 0,
              "touchedFields": Object {},
            },
          },
          "getIsDirty": [Function],
          "inFieldArrayActionRef": Object {
            "current": false,
          },
          "namesRef": Object {
            "current": Object {
              "array": Set {},
              "mount": Set {},
              "unMount": Set {},
              "watch": Set {},
              "watchAll": false,
            },
          },
          "readFormStateRef": Object {
            "current": Object {
              "dirtyFields": false,
              "errors": "all",
              "isDirty": false,
              "isValid": false,
              "isValidating": false,
              "touchedFields": false,
            },
          },
          "register": [Function],
          "shouldUnmount": undefined,
          "subjectsRef": Object {
            "current": Object {
              "array": ke {
                "observers": Array [],
              },
              "control": ke {
                "observers": Array [],
              },
              "state": ke {
                "observers": Array [],
              },
              "watch": ke {
                "observers": Array [],
              },
            },
          },
          "unregister": [Function],
          "updateIsValid": [Function],
          "watchInternal": [Function],
        }
      }
      errors={Object {}}
    />
  </Dialog>
</div>
`;

exports[`START: Enable button 1`] = `
<div
  className="Tokenization"
>
  <Button
    onClick={[Function]}
  >
    Request token creation
  </Button>
  <Dialog
    actions={
      Array [
        Object {
          "buttonText": "Submit",
          "buttonVariant": "primary",
          "id": "submit",
          "type": "submit",
        },
        Object {
          "buttonText": "Cancel",
          "buttonVariant": "secondary",
          "callback": [Function],
          "id": "cancel",
        },
      ]
    }
    onSubmit={[Function]}
    show={false}
    size="lg"
  >
    <TokenCreationForm
      colors={
        Object {
          "background": "#eff3fe",
          "borderColor": "#e6007a",
          "foreground": "#000000",
        }
      }
      control={
        Object {
          "defaultValuesRef": Object {
            "current": Object {},
          },
          "fieldArrayDefaultValuesRef": Object {
            "current": Object {},
          },
          "fieldsRef": Object {
            "current": Object {},
          },
          "formStateRef": Object {
            "current": Object {
              "dirtyFields": Object {},
              "errors": Object {},
              "isDirty": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": false,
              "isValidating": false,
              "submitCount": 0,
              "touchedFields": Object {},
            },
          },
          "getIsDirty": [Function],
          "inFieldArrayActionRef": Object {
            "current": false,
          },
          "namesRef": Object {
            "current": Object {
              "array": Set {},
              "mount": Set {},
              "unMount": Set {},
              "watch": Set {},
              "watchAll": false,
            },
          },
          "readFormStateRef": Object {
            "current": Object {
              "dirtyFields": false,
              "errors": "all",
              "isDirty": false,
              "isValid": false,
              "isValidating": false,
              "touchedFields": false,
            },
          },
          "register": [Function],
          "shouldUnmount": undefined,
          "subjectsRef": Object {
            "current": Object {
              "array": ke {
                "observers": Array [],
              },
              "control": ke {
                "observers": Array [],
              },
              "state": ke {
                "observers": Array [],
              },
              "watch": ke {
                "observers": Array [],
              },
            },
          },
          "unregister": [Function],
          "updateIsValid": [Function],
          "watchInternal": [Function],
        }
      }
      errors={Object {}}
    />
  </Dialog>
</div>
`;
